File: src/main/scala/org/xbery/artbeams/orders/service/OrderService.scala
package org.xbery.artbeams.orders.service

import java.time.Instant

import org.xbery.artbeams.orders.domain.{Order, OrderItem}

/**
  * @author Radek Beran
  */
trait OrderService {

  def createOrder(order: Order): Order

  /**
    * Finds order item representing an order of given product by given user.
    * @param userId
    * @param productId
    * @return
    */
  def findOrderItemOfUser(userId: String, productId: String): Option[OrderItem]

  /**
    * Updates downloaded time of order item.
    * @param orderItemId
    * @param downloaded downloaded time
    * @return
    */
  def updateOrderItemDownloaded(orderItemId: String, downloaded: Option[Instant]): Option[Instant]
}

File: src/main/kotlin/org/xbery/artbeams/orders/service/OrderService.kt
package org.xbery.artbeams.orders.service

import org.xbery.artbeams.orders.domain.Order
import org.xbery.artbeams.orders.domain.OrderInfo
import org.xbery.artbeams.orders.domain.OrderItem
import org.xbery.artbeams.orders.domain.OrderState
import org.xbery.artbeams.products.domain.Product
import java.time.Instant

/**
 * @author Radek Beran
 */
interface OrderService {
    fun generateOrderNumber(): String

    fun createOrderOfProduct(userId: String, product: Product): Order

    fun createOrderOfProduct(userId: String, product: Product, orderNumber: String, orderState: OrderState): Order

    fun createOrder(order: Order): Order

    fun findOrders(): List<OrderInfo>

    fun requireByOrderNumber(orderNumber: String): Order

    fun requireByOrderId(orderId: String): Order

    fun updateOrderPaid(orderId: String): Boolean

    fun updateOrderState(orderId: String, state: OrderState): Boolean

    fun deleteOrder(orderId: String): Boolean

    fun findOrderItemsOfUserAndProduct(userId: String, productId: String): List<OrderItem>

    /**
     * Updates downloaded time of order item.
     * @param orderItemId
     * @param downloaded downloaded time
     * @return
     */
    fun updateOrderItemDownloaded(orderItemId: String, downloaded: Instant?): OrderItem
}
